thorlcr/master/thgraphmanager.cpp

```diff
@@ -1456,63 +1457,153 @@ void thorMain(ILogMsgHandler *logHandler, const char *wuid, const char *graphNam
                         }
                         else
                         {
-                            JobNameScope activeJobName(currentWuid.str());
-                            saveWuidToFile(currentWuid);
-                            PROGLOG("Executing: wuid=%s, graph=%s", currentWuid.str(), currentGraphName.str());
-
+                            SessionId agentSessionID = workunit->getAgentSession();
+                            if (agentSessionID <= 0)
                             {
-                                Owned<IWorkUnit> wu = &workunit->lock();
-                                publishPodNames(wu, currentGraphName);
+                                WARNLOG("Discarding job with invalid sessionID: wuid=%s, graph=%s (sessionID=%" I64F "d)", currentWuid.str(), currentGraphName.str(), agentSessionID);
+                                currentWuid.clear();
                             }
-                            SocketEndpoint dummyAgentEp;
-                            jobManager->execute(workunit, currentWuid, currentGraphName, dummyAgentEp);
-
-                            Owned<IWorkUnit> w = &workunit->lock();
-                            if (!multiJobLinger && lingerPeriod)
-                                w->setDebugValue(instance, "1", true);
-
-                            if (jobManager->queryExitException())
+                            else if (querySessionManager().sessionStopped(agentSessionID, 0))
                             {
-                                // NB: exitException has already been relayed.
-                                jobManager->clearExitException();
+                                WARNLOG("Discarding agentless job: wuid=%s, graph=%s", currentWuid.str(), currentGraphName.str());
+                                currentWuid.clear();
                             }
                             else
                             {
-                                switch (w->getState())
+                                SCMStringBuffer jobVersion;
+                                workunit->getDebugValue("platformVersion", jobVersion);
+                                if (jobVersion.length() && !thorQueue)
                                 {
-                                    case WUStateRunning:
-                                        w->setState(WUStateWait);
-                                        break;
-                                    case WUStateAborting:
-                                    case WUStateAborted:
-                                    case WUStateFailed:
-                                        break;
-                                    default:
-                                        w->setState(WUStateFailed);
-                                        break;
+                                    Owned<IException> exception = MakeStringException(TE_AbortException, "Custom platformVersion with multiJobLinger=false not supported");
```

Comment:
Why does the non-linger case not "just work"?

Reply:
yes, though I discovered an issue which has broken a configuration when muiltiJobLinger is false.
(the change to persist hostnames in endpoints broke it)